### PR TITLE
fix: Sentry HTTP Trace Header Breaking Requests

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Removed SentryURLProtocol that was responsible for adding trace header in every request. The `startLoading`
used `NSURLSessionConfiguration.defaultSessionConfiguration`, which didn't include custom protocol classes. 
This is fixed now by swizzling `NSURLSessionConfiguration.HTTPAdditionalHeaders`. Whenever this getter is
called, the SDK adds the trace header if there is a span on the scope.

## :bulb: Motivation and Context

Fixes GH-1288

## :green_heart: How did you test it?
Unit tests, Integration tests, simulators, and sending a distributed trace to a local running Java server. For reference see this branch: https://github.com/getsentry/sentry-cocoa/compare/workshop?expand=1

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by bruno-garcia** [comment]
Looking

File: Sources/Sentry/SentryNetworkTracker.m [hunk 1]
```diff
@@ -252,4 +245,20 @@ - (SentrySpanStatus)spanStatusForHttpResponseStatusCode:(NSInteger)statusCode
...
 
+- (nullable NSDictionary *)addTraceHeader:(nullable NSDictionary *)headers
+{
+    id<SentrySpan> span = SentrySDK.currentHub.scope.span;
+    if (span == nil) {
+        return headers;
+    }
+
+    if (headers == nil) {
+        return @{ SENTRY_TRACE_HEADER : [span toTraceHeader].value };
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryNetworkTracker.m:256]: `[span toTraceHeader].value` guaranteed to return something that makes sense?
Or do we need to validate that before adding the header.

File: Sources/Sentry/SentryNetworkTracker.m [hunk 2]
```diff
@@ -252,4 +245,20 @@ - (SentrySpanStatus)spanStatusForHttpResponseStatusCode:(NSInteger)statusCode
     return kSentrySpanStatusUndefined;
 }
 
+- (nullable NSDictionary *)addTraceHeader:(nullable NSDictionary *)headers
+{
+    id<SentrySpan> span = SentrySDK.currentHub.scope.span;
+    if (span == nil) {
+        return headers;
+    }
+
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryNetworkTracker.m:254]: Perhaps assign the resulting dictionary to a variable and after the `if` you could do:

If we have a check for the log level, we can do this:
```
if (SentrySDK.logger.isEnabled(SentryLevel.Debug)) {
// Expensive operations done just for logging go here
}
```

**Review by bruno-garcia — APPROVED** [review]

File: Sources/Sentry/SentryNetworkTracker.m (hunk 1 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryNetworkTracker.m:256]: It's guaranteed to return something meaningful as `toTraceHeader` is `nonnull` and also `.value`.

**Review by brustolin — COMMENTED** [review]
LGTM

In the future, to support NSURLConnection we could think about doing the same thing with NSURLRequest.allHTTPHeaderFields instead of SessionConfiguration.addicionalHeaders.

**Comment by philipphofmann** [comment]
> In the future, to support NSURLConnection we could think about doing the same thing with NSURLRequest.allHTTPHeaderFields instead of SessionConfiguration.addicionalHeaders.

I already have some stuff here https://github.com/getsentry/sentry-cocoa/tree/feat/trace-header-for-nsurlconnection.

File: Sources/Sentry/SentryNetworkTracker.m (hunk 2 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryNetworkTracker.m:254]: Will add that in the future.

**Pull request merged**